File: tests/vectors/ber_primitives/test_ber_primitives.py
"""Runtime BER primitive tests using YAML test vectors."""
import os
import sys
from pathlib import Path
import yaml
import pytest

from asnvil_runtime import Tag, TagClass, BerDecoder

# Load the test vectors from the co-located data.yaml file
VECTORS_PATH = Path(__file__).parent / "data.yaml"
with open(VECTORS_PATH) as f:
    VECTORS = yaml.safe_load(f)

# Filter to only include vectors that test tag decoding (have 'expected' with tag_class)
TAG_VECTORS = [v for v in VECTORS if v.get("expected", {}).get("tag_class") is not None]


@pytest.mark.parametrize("vector", TAG_VECTORS, ids=lambda v: v["name"])
def test_ber_tag_decoding(vector):
    """Test BER tag decoding against known vectors."""
    ber_bytes = bytes.fromhex(vector["ber_hex"])
    expected = vector["expected"]
    
    decoder = BerDecoder(ber_bytes)
    tag_class, tag_number, constructed = decoder.read_tag()
    
    # Map string tag class to enum
    expected_class_map = {
        "universal": TagClass.UNIVERSAL,
        "application": TagClass.APPLICATION,
        "context": TagClass.CONTEXT_SPECIFIC,
        "private": TagClass.PRIVATE,
    }
    
    assert tag_class == expected_class_map[expected["tag_class"].lower()], \
        f"Tag class mismatch for {vector['name']}"
    assert tag_number == expected["tag_number"], \
        f"Tag number mismatch for {vector['name']}"
    assert constructed == expected["constructed"], \
        f"Constructed flag mismatch for {vector['name']}"

File: tests/vectors/ber_primitives/test_ber_primitives.rs
//! ```cargo
//! [dependencies]
//! asnvil-runtime-rust = { path = "__REPO_ROOT__/asnvil-runtime-rust" }
//! ```

use asnvil_runtime_rust::{TagClass, BerDecoder};

fn main() {
    test_ber_tag_decoding_universal();
    test_ber_tag_decoding_context();
    test_ber_tag_decoding_application();
    test_ber_tag_decoding_private();
    test_ber_tag_decoding_integer();
    println!("\nAll BER Primitives integration tests passed!");
}

fn test_ber_tag_decoding_universal() {
    // UNIVERSAL 2 (INTEGER)
    let ber_bytes = [0x02, 0x01, 0x05];
    let mut decoder = BerDecoder::new(&ber_bytes);
    let (tag_class, tag_number, constructed) = decoder.read_tag().expect("Failed to read tag");
    assert_eq!(tag_class, TagClass::Universal);
    assert_eq!(tag_number, 2);
    assert!(!constructed);
    println!("PASS: test_ber_tag_decoding_universal");
}

fn test_ber_tag_decoding_context() {
    // CONTEXT 0
    let ber_bytes = [0x80, 0x00];
    let mut decoder = BerDecoder::new(&ber_bytes);
    let (tag_class, tag_number, constructed) = decoder.read_tag().expect("Failed to read tag");
    assert_eq!(tag_class, TagClass::Context);
    assert_eq!(tag_number, 0);
    assert!(!constructed);
    println!("PASS: test_ber_tag_decoding_context");
}

fn test_ber_tag_decoding_application() {
    // APPLICATION 3
    let ber_bytes = [0x63, 0x00];
    let mut decoder = BerDecoder::new(&ber_bytes);
    let (tag_class, tag_number, constructed) = decoder.read_tag().expect("Failed to read tag");
    assert_eq!(tag_class, TagClass::Application);
    assert_eq!(tag_number, 3);
    assert!(!constructed);
    println!("PASS: test_ber_tag_decoding_application");
}

fn test_ber_tag_decoding_private() {
    // PRIVATE 27
    let ber_bytes = [0xDF, 0x1B, 0x00];
    let mut decoder = BerDecoder::new(&ber_bytes);
    let (tag_class, tag_number, constructed) = decoder.read_tag().expect("Failed to read tag");
    assert_eq!(tag_class, TagClass::Private);
    assert_eq!(tag_number, 27);
    assert!(!constructed);
    println!("PASS: test_ber_tag_decoding_private");
}

fn test_ber_tag_decoding_integer() {
    // INTEGER 1
    let ber_bytes = [0x02, 0x01, 0x01];
    let mut decoder = BerDecoder::new(&ber_bytes);
    let (tag_class, tag_number, constructed) = decoder.read_tag().expect("Failed to read tag");
    assert_eq!(tag_class, TagClass::Universal);
    assert_eq!(tag_number, 2);
    assert!(!constructed);
    let val = decoder.read_integer().expect("Failed to read integer");
    assert_eq!(val, 1);
    println!("PASS: test_ber_tag_decoding_integer");
}